include/ensmallen_bits/adamr/adamr_impl.hpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file adamr_impl.hpp
+ * @author Niteya Shah
+ * Implementation of the AdamR optimizer.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_ADAM_ADAMR_IMPL_HPP
+#define ENSMALLEN_ADAM_ADAMR_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "adamr.hpp"
+
+namespace ens {
+
+template<typename UpdateRule>
+AdamRType<UpdateRule>::AdamRType(
+    const double stepSize,
+    const size_t epochRestart,
+    const double multFactor,
+    const size_t batchSize,
+    const double beta1,
+    const double beta2,
+    const double epsilon,
+    const size_t maxIterations,
+    const double tolerance,
+    const bool shuffle,
+    const bool resetPolicy) :
+    batchSize(batchSize),
+    optimizer(stepSize,
+              batchSize,
+              maxIterations,
+              tolerance,
+              shuffle,
+              UpdateRule(epsilon, beta1, beta2),
+              CyclicalDecay(
+                  epochRestart,
+                  multFactor,
+                  stepSize),
+              resetPolicy)
+{ /* Nothing to do. */ }
+
+template<typename UpdateRule>
+AdamRType<UpdateRule>::AdamRType(
+    const double stepSize,
+    const double stepSizeMin,
+    const size_t epochRestart,
+    const double multFactor,
+    const size_t batchSize,
+    const double beta1,
+    const double beta2,
+    const double epsilon,
+    const size_t maxIterations,
+    const double tolerance,
+    const bool shuffle,
+    const bool resetPolicy) :
+    batchSize(batchSize),
+    optimizer(stepSize,
+              batchSize,
+              maxIterations,
+              tolerance,
+              shuffle,
+              UpdateRule(epsilon, beta1, beta2),
+              CyclicalDecay(
+                  epochRestart,
+                  multFactor,
+                  stepSize,
+                  stepSizeMin),
+              resetPolicy)
+{ /* Nothing to do. */ }
+
+template<typename UpdatePolicyType>
+template<typename DecomposableFunctionType>
+double AdamRType<UpdatePolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // If a user changed the step size he hasn't update the step size of the
+  // cyclical decay instantiation, so we have to do it here.
+  if (optimizer.StepSize() != optimizer.DecayPolicy().StepSize())
+  {
+    optimizer.DecayPolicy().StepSize() = optimizer.StepSize();
+  }
+
+  optimizer.DecayPolicy().EpochBatches() = function.NumFunctions() /
+      double(optimizer.BatchSize());
```

Comment:
@zoq Can you please tell me what this  line does ?

Reply:
It describes the decay value per number of function evaluations per batch. Here is an example, let say we like to train a neural network with 10 samples, and a batch size of 2, in this case ``function.NumFunctions()`` returns 10 and ``optimizer.BatchSize()`` returns 2.

Reply:
In Cyclical Decay , isnt the decay value set by function itself ? Why are we manually setting it ?